src/prefect/utilities/executors.py

```diff
@@ -39,52 +54,151 @@ def run_with_heartbeat(
+        except Exception:
+            use_heartbeat = False
+            logger = self.logger
+            logger.exception(
+                "Heartbeat process is misconfigured.  This could result in a zombie run.",
+                exc_info=True,
+            )
+
+        if not use_heartbeat or (prefect.context.config.heartbeat_mode == "off"):
+            configured_heartbeat = no_heartbeat()
+        elif prefect.context.config.heartbeat_mode == "thread":
+            configured_heartbeat = threaded_heartbeat(self.flow_run_id)
+        elif prefect.context.config.heartbeat_mode == "process":
+            configured_heartbeat = subprocess_heartbeat(self.heartbeat_cmd, self.logger)
+        # because the threaded heartbeat mode is experimental and may change in the future,
+        # let's not catch configuration error with `else` -- stale configuration should break tests
+
+        with configured_heartbeat:
             return runner_method(self, *args, **kwargs)
-        finally:
-            if p is not None:
-                exit_code = p.poll()
-                if exit_code is not None:
-                    msg = "Heartbeat process died with exit code {}".format(exit_code)
-                    self.logger.error(msg)
-                p.kill()
-                p.wait()
 
     return inner
 
 
+@contextlib.contextmanager
+def no_heartbeat() -> Iterator[None]:
+    # contextlib.nullcontext was introduced in 3.7
+    yield
+
+
+@contextlib.contextmanager
+def threaded_heartbeat(flow_run_id: str, num: int = None) -> Iterator[None]:
+    try:
+        HEARTBEAT_STOP_EVENT = threading.Event()
+        heartbeat = HeartbeatThread(HEARTBEAT_STOP_EVENT, flow_run_id, num=None)
+        heartbeat.start()
+        yield
+    finally:
+        HEARTBEAT_STOP_EVENT.set()
+
+
+@contextlib.contextmanager
+def subprocess_heartbeat(heartbeat_cmd: List[str], logger: Logger) -> Iterator[None]:
+    p = None
+    try:
+        # we use Popen + a prefect CLI for a few reasons:
+        # - using threads would interfere with the task; for example, a task
+        #   which does not release the GIL would prevent the heartbeat thread from
+        #   firing
+        # - using multiprocessing.Process would release the GIL but a subprocess
+        #   cannot be spawned from a daemonic subprocess, and Dask sometimes will
+        #   submit tasks to run within daemonic subprocesses
+        current_env = dict(os.environ).copy()
+        auth_token = prefect.context.config.cloud.get("auth_token")
+        api_key = prefect.context.config.cloud.get("api_key")
+        tenant_id = prefect.context.config.cloud.get("tenant_id")
+        api_url = prefect.context.config.cloud.get("api")
+        current_env.setdefault("PREFECT__CLOUD__AUTH_TOKEN", auth_token)
+        current_env.setdefault("PREFECT__CLOUD__API_KEY", api_key)
+        current_env.setdefault("PREFECT__CLOUD__TENANT_ID", tenant_id)
+        current_env.setdefault("PREFECT__CLOUD__API", api_url)
```

Comment:
There is a utility that make this cleaner at `prefect.utilities.configuration.to_environment_variables`

Reply:
e.g.
```python
    env.update(
        to_environment_variables(
            prefect.context.config,
            include={
                "logging.level",
                "logging.format",
                "logging.datefmt",
                "cloud.send_flow_run_logs",
            },
        )
    )
```